### PR TITLE
ClassToAPI.classCanonicalName spec

### PR DESCRIPTION
**Pull request by dwijnand**

### CODE REVIEW

File: internal/zinc-apiinfo/src/test/scala/sbt/internal/inc/ClassCanonicalNameSpec.scala [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+  data foreach { case Data(nesting, full, refClass, tagClass, canon) =>
+    it should f"for $nesting return $full" in assert(
+      run(refClass) === full &&
+      run(tagClass) === full &&
+      run2(refClass) === canon &&
+      run2(tagClass) === canon
+    )
+  }
+
+  // O means object, C means cleans, therefore..
```
Comment (jvican) [inline internal/zinc-apiinfo/src/test/scala/sbt/internal/inc/ClassCanonicalNameSpec.scala]: typo: class

File: internal/zinc-apiinfo/src/test/scala/sbt/internal/inc/ClassCanonicalNameSpec.scala [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    C.create[p3.x](new p3.x),
+
+    OO.create[p1.x.y.type](p1.x.y),
+    OC.create[p2.x.y](new p2.x.y),
+    CO.create[c1.y.type forSome { val c1: p3.x }]({ val c1 = new p3.x ; c1.y }),
+    CC.create[p4.x#y]({ val c1 = new p4.x ; new c1.y }),
+
+    OOO.create[p1.x.y.z.type](p1.x.y.z),
+    OOC.create[p1.x.y.z](new p1.x.y.z),
+    OCO.create[c2.z.type forSome { val c2: p2.x.y }]({ val c2 = new p2.x.y ; c2.z }),
```
Comment (jvican) [inline internal/zinc-apiinfo/src/test/scala/sbt/internal/inc/ClassCanonicalNameSpec.scala]: :+1:

File: internal/zinc-apiinfo/src/test/scala/sbt/internal/inc/ClassCanonicalNameSpec.scala [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    }
+  }
+
+  case class Data(nesting: String, full: String, refClass: Class[_], tagClass: Class[_], canon: String)
+}
+
+// All eight combinations of class/object nesting to three levels.
+// Because of scala/bug#2034 we can't put these all in one package (you get "name clash" errors)
+// So instead we'll split them in 4 nice and even packages
+package p1 { object x { object y { object z; class z } } }
```
Comment (jvican) [inline internal/zinc-apiinfo/src/test/scala/sbt/internal/inc/ClassCanonicalNameSpec.scala]: Does this pass formatting?

**Review by jvican — DISMISSED** [review]
LGTM, minor question. Good tests.

File: internal/zinc-apiinfo/src/test/scala/sbt/internal/inc/ClassCanonicalNameSpec.scala [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+  object CC extends Expected("x.y")
+
+  object OOO extends Expected("x.y$$z$", nativeClassNameIsMalformed = true)
+  object OOC extends Expected("x.y$$z", nativeClassNameIsMalformed = true)
+  object OCO extends Expected("x.y.z$")
+  object OCC extends Expected("x.y.z")
+  object COO extends Expected("x.y$$z$", nativeClassNameIsMalformed = true)
+  object COC extends Expected("x.y$$z", nativeClassNameIsMalformed = true)
+  object CCO extends Expected("x.y.z$")
+  object CCC extends Expected("x.y.z")
```
Comment (dwijnand) [inline internal/zinc-apiinfo/src/test/scala/sbt/internal/inc/ClassCanonicalNameSpec.scala:90]: @jvican Do these canonical class names look right to you?

Looking at the pattern across all the nestings, the ones that have a malformed class name look wrong to me.

Basically I expect them all to be either "x.y.z$" or "x.y.z".

WDYT?

File: internal/zinc-apiinfo/src/test/scala/sbt/internal/inc/ClassCanonicalNameSpec.scala (hunk 3 above)
Comment (dwijnand) [reply internal/zinc-apiinfo/src/test/scala/sbt/internal/inc/ClassCanonicalNameSpec.scala]: Not sure. We don't have neo-sbt-scalafmt ATM, and I don't have an up-to-date CLI launcher. Let me know what it should look like and I'll change it; meanwhile I've dropped the space tabbing I had... 😢

File: internal/zinc-apiinfo/src/test/scala/sbt/internal/inc/ClassCanonicalNameSpec.scala (hunk 4 above)
Comment (jvican) [reply internal/zinc-apiinfo/src/test/scala/sbt/internal/inc/ClassCanonicalNameSpec.scala:90]: I've answered this in #381.

**Comment by dwijnand** [comment]
Let's keep this one open in case we decide to not change the behaviour in #381, we still merge the spec.

**Review by jvican — DISMISSED** [review]
LGTM.

**Comment by jvican** [comment]
This makes sense. As there's no change in behaviour in this one, I'm merging it.

**Review by jvican — APPROVED** [review]

**Pull request merged**